editoast/src/views/timetable/simulation.rs

```diff
@@ -60,6 +63,125 @@ pub struct SimulationResponseSuccess {
+                continue;
+            };
+            let arrival = Duration::from(arrival);
+
+            let path_item_index = path_items_id2index[&schedule_item.at];
+            let path_item_time = i64::try_from(path_item_times_final[path_item_index]).unwrap();
+            let path_item_time = Duration::milliseconds(path_item_time);
+
+            res[path_item_index] = (path_item_time - arrival).abs() < Duration::seconds(2);
+        }
+
+        res
+    }
+
+    /// Compute whether the simulation response respects the margins on each path item.
+    ///
+    /// # Panics
+    ///
+    /// This function can panic in the following cases:
+    /// - the simulation response isn't derived from the given train schedule,
+    /// - some of the final or provisional path item times exceed `i64::MAX`
+    pub fn path_items_respect_margins<T: TrainScheduleLike>(
+        &self,
+        train_schedule: &T,
+    ) -> Vec<bool> {
+        let path_item_times_final = &*self.final_output.report_train.path_item_times;
+        let path_item_times_provisional = &*self.provisional.path_item_times;
+        let margin_boundary_set = &*train_schedule.margins().boundaries;
+
+        let path_item_id2index: HashMap<&NonBlankString, usize> = train_schedule
+            .path()
+            .iter()
+            .enumerate()
+            .map(|(i, path_item)| (&path_item.id, i))
+            .collect();
+
+        let mut res = vec![true; path_item_times_final.len()];
+
+        let path_item_ids_to_check = train_schedule
+            .path()
+            .first()
+            .map(|first_path_item| &first_path_item.id)
+            .into_iter()
+            .chain(
+                train_schedule
+                    .schedule()
+                    .iter()
+                    .enumerate()
+                    .filter(|(i, schedule_item)| {
+                        *i == 0
+                            || schedule_item.arrival.is_some()
+                            || margin_boundary_set.contains(&schedule_item.at)
+                    })
+                    .map(|(_, schedule_item)| &schedule_item.at),
+            )
+            .chain(
+                train_schedule
+                    .path()
+                    .last()
+                    .map(|last_path_item| &last_path_item.id),
+            )
+            // TODO use https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.map_windows
+            .collect::<Vec<&NonBlankString>>();
+        path_item_ids_to_check
+            .windows(2)
+            .for_each(|path_item_id_couple| {
+                let [prev_path_item_id, path_item_id] = *path_item_id_couple else {
+                    // TODO use https://doc.rust-lang.org/stable/std/primitive.slice.html#method.array_windows
+                    unreachable!()
+                };
+                if prev_path_item_id == path_item_id {
+                    return;
+                }
```

Comment:
I'm guessing this is how the potential duplication of the first and last path items is handled? If so a small comment in the iterator construction and here could help understand that (I actually removed about comment about that above after reading this condition).